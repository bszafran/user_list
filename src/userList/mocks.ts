import { User } from "./model"

const mockedUsers: User[] = [
  {
    id: 1,
    name: "Janusz Kowalski",
    username: "januszk",
  },
  {
    id: 3,
    name: "Micha≈Ç Nowak",
    username: "nowakm",
  },
]

export { mockedUsers }
