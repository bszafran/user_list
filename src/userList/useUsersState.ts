import { useEffect, useState } from "react"
import { fetchUsers } from "./api"
import { User } from "./model"

const useUsersState = () => {
  const [users, setUsers] = useState<User[]>([])
  const [filteredUsers, setFilteredUsers] = useState<User[]>([])
  const [search, setSearch] = useState<string>("")
  const [error, setError] = useState<boolean>(false)

  useEffect(() => {
    fetchUsers()
      .then((data) => setUsers(data))
      .catch(() => setError(true))
  }, [])

  useEffect(() => {
    const lowerCasedSearch = search.toLowerCase()

    setFilteredUsers(
      users.filter((user) => {
        const words = user.name.toLowerCase().split(" ")
        return words.some((word) => word.startsWith(lowerCasedSearch))
      })
    )
  }, [search, users])

  return {
    users: filteredUsers,
    search,
    setSearch,
    error,
  }
}

export default useUsersState
