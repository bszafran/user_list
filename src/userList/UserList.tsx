import React from "react"
import styled from "styled-components/macro"
import Input from "../components/Input"
import useUsersState from "./useUsersState"
import UserEntry from "./UserEntry"

const UserListContainer = styled.div`
  display: flex;
  flex-direction: column;
`

const ListContainer = styled.div`
  min-width: 350px;

  display: flex;
  flex-direction: column;

  margin-top: 30px;
  margin-left: 5px;
`

const InputContainer = styled.div`
  margin-top: 10px;

  input {
    width: 250px;
  }
`

const Header = styled.h1`
  margin-left: 50px;
`

const UserList = () => {
  const { users, search, setSearch, error } = useUsersState()

  return (
    <UserListContainer>
      <Header>Users list</Header>
      <InputContainer>
        <Input
          value={search}
          onChange={setSearch}
          placeholder="Search by user name..."
        />
      </InputContainer>
      <ListContainer>
        {error ? (
          <div>We are sorry, data cannot be loaded.</div>
        ) : (
          users.map((user, index) => (
            <UserEntry key={user.id} orderNumber={index + 1} user={user} />
          ))
        )}
      </ListContainer>
    </UserListContainer>
  )
}

export default UserList
