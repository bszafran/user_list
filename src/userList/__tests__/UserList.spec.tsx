import { shallow } from "enzyme"
import React from "react"
import { mockedUsers } from "../mocks"
import UserList from "../UserList"
import useUsersState from "../useUsersState"

jest.mock("../useUsersState")
const mockedUseUsersState = useUsersState as jest.Mock

describe("UserList component", () => {
  const mockedUseUsersStateValue = {
    users: mockedUsers,
    search: "",
    setSearch: () => {},
    error: false,
  }

  it("displays data when available", async () => {
    mockedUseUsersState.mockReturnValue(mockedUseUsersStateValue)
    const wrapper = shallow(<UserList />)

    expect(wrapper).toMatchSnapshot()
  })

  it("renders empty list container when no data are available", () => {
    mockedUseUsersState.mockReturnValue({
      ...mockedUseUsersStateValue,
      users: [],
    })
    const wrapper = shallow(<UserList />)

    expect(wrapper).toMatchSnapshot()
  })

  it("displays error message when error occurred", () => {
    mockedUseUsersState.mockReturnValue({
      ...mockedUseUsersStateValue,
      error: true,
    })
    const wrapper = shallow(<UserList />)

    expect(wrapper).toMatchSnapshot()
  })
})
