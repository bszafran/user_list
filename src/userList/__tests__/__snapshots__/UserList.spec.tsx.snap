// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList component displays data when available 1`] = `
<UserList__UserListContainer>
  <UserList__Header>
    Users list
  </UserList__Header>
  <UserList__InputContainer>
    <Input
      onChange={[Function]}
      placeholder="Search by user name..."
      value=""
    />
  </UserList__InputContainer>
  <UserList__ListContainer>
    <UserEntry
      key="1"
      orderNumber={1}
      user={
        Object {
          "id": 1,
          "name": "Janusz Kowalski",
          "username": "januszk",
        }
      }
    />
    <UserEntry
      key="3"
      orderNumber={2}
      user={
        Object {
          "id": 3,
          "name": "MichaÅ‚ Nowak",
          "username": "nowakm",
        }
      }
    />
  </UserList__ListContainer>
</UserList__UserListContainer>
`;

exports[`UserList component displays error message when error occurred 1`] = `
<UserList__UserListContainer>
  <UserList__Header>
    Users list
  </UserList__Header>
  <UserList__InputContainer>
    <Input
      onChange={[Function]}
      placeholder="Search by user name..."
      value=""
    />
  </UserList__InputContainer>
  <UserList__ListContainer>
    <div>
      We are sorry, data cannot be loaded.
    </div>
  </UserList__ListContainer>
</UserList__UserListContainer>
`;

exports[`UserList component renders empty list container when no data are available 1`] = `
<UserList__UserListContainer>
  <UserList__Header>
    Users list
  </UserList__Header>
  <UserList__InputContainer>
    <Input
      onChange={[Function]}
      placeholder="Search by user name..."
      value=""
    />
  </UserList__InputContainer>
  <UserList__ListContainer />
</UserList__UserListContainer>
`;
